#!/usr/bin/env Rscript

library(optparse)

option_list <- list(
    make_option(c("-p", "--prefix"), type = "character", dest = "prefix",
                help = "prefix of input and output files (required)", metavar = "STRING"),
    make_option(c("-o", "--out"), type = "character", dest = "dir_output",
                help = "path to output directory (required)", metavar = "DIR"),
    make_option(c("-d", "--derep"), type = "character", dest = "dir_derep",
                help = "path to derep directory (generated by derep_fastq.R) (required)", metavar = "DIR"),
    make_option(c("-b", "--blast"), type = "character", dest = "dir_blast",
                help = "path to blast directory (generated by derep_fastq.R) (required)", metavar = "DIR"),
    make_option(c("-m", "--metadata"), type = "character", dest = "excel_meta",
                help = "path to TSV of Excel file of metadata", metavar = "FILE"),
    ## make_option(c("-c", "--gene-column"), type = "character", dest = "gene_column",
    ##             help = "name of column containing gene name; exclude ' Set X' suffix"),
    make_option(c("-g", "--gene-pattern"), type = "character", dest = "gene_pattern", default = "[^_]+$",
                help = "gene regex; used with str_extract to get gene ID from Excel file")
)
parser <- OptionParser(option_list = option_list)
args <- parse_args(parser)

## check argument
if (!is.na(args$prefix)){
    prefix <- args$prefix
} else {
    stop("Prefix must be provided. Use '--prefix <IN/OUT PREFIX>'.")
}
if (is.na(args$dir_output)){
    stop("Output directory must be provided. Use '--out <DIR>'.")
}
if (!is.na(args$dir_derep)){
    dir_derep <- normalizePath(args$dir_derep)
} else {
    stop("Derep directory must be provided. Use '--derep <PATH>'.")
}
if (!is.na(args$dir_blast)){
    dir_blast <- normalizePath(args$dir_blast)
} else {
    stop("BLAST directory must be provided. Use '--blast <PATH>'.")
}
if (!is.na(args$excel_meta)){
    excel_meta <- normalizePath(args$excel_meta)
} else {
    stop("TSV file of metadata must be provided. Use '--metadata <PATH>'.")
}

suppressMessages(library(tidyverse))

mkpath <- function(...){
    paste(..., sep = '/')
}

parse_blast <- function(fname){
    read.table(fname, header = TRUE, sep = '\t') %>%
        dplyr::group_by(qseqid) %>%
        dplyr::slice_max(bitscore, n = 1) %>%
        dplyr::mutate(gene = as.character(gene),
                      gchrom = as.character(gchrom))
}

parse_meta <- function(fname, Nmultiplex = 6){
    df.raw <- read.table(fname, header = TRUE, sep = '\t', quote = '', comment.char = '')
    df <- data.frame()
    common_cols <- c("fastq", "Index1", "Index2")
    uniq_cols <- c("user", "accession.gene", "PCR_Product_fa", "Guide")
    for (i in 1:Nmultiplex){
        cols <- paste0(uniq_cols, ".Set", i)
        df.tmp <- df.raw[names(df.raw) %in% c(common_cols, cols)]
        for (j in 1:length(uniq_cols)){
            colnames(df.tmp)[j + length(common_cols)] <- uniq_cols[[j]]
        }
        df <- rbind(df, df.tmp)
    }
    return(df %>% dplyr::filter(!is.na(user) & user != "Â " & user != ""))
}

summarise_reads <- function(sample_id, df.meta, dir_summary, dir_blast, valid.genes = c()){
    ## names
    fastq_id <- str_split(sample_id, pattern = "_")[[1]][[1]]
    fname_sum <- mkpath(dir_summary, paste0(sample_id, ".sum.txt"))
    fname_fwd <- mkpath(dir_blast, paste0(sample_id, ".fwd.blastn.intersectGene.tsv"))
    fname_rvs <- mkpath(dir_blast, paste0(sample_id, ".rvs.blastn.intersectGene.tsv"))
    ## parse files
    df.sum <- read.table(fname_sum, header = TRUE, sep = '\t')
    df.fwd <- parse_blast(fname_fwd) %>%
        dplyr::select(qseqid, sstart, send, gene, gchrom, bitscore) %>%
        dplyr::rename(fwd = qseqid, fstart = sstart, fend = send, fgene = gene, fchrom = gchrom,
                      fbitscore = bitscore)
    df.rvs <- parse_blast(fname_rvs) %>%
        dplyr::select(qseqid, sstart, send, gene, gchrom, bitscore) %>%
        dplyr::rename(rvs = qseqid, rstart = sstart, rend = send, rgene = gene, rchrom = gchrom,
                      rbitscore = bitscore)
    ## join fwd and rvs to read count summary file
    df.sum.full <- df.sum %>%
        dplyr::left_join(df.fwd, by = "fwd", relationship = "many-to-many") %>%
        dplyr::left_join(df.rvs, by = "rvs", relationship = "many-to-many") %>%
        dplyr::mutate(same.gene = fgene == rgene,
                      within.range = (fchrom == rchrom &
                                      abs(((fstart + fend)/2) - ((rstart + rend)/2)) < 400))
    ## get 
    total.reads <- df.sum %>% dplyr::pull(reads) %>% sum
    ## merge df.meta genes with valid.genes if df.meta is not NA
    if (!is.null(df.meta)){
        valid.genes <- c(valid.genes,
                         df.meta %>%
                         dplyr::filter(fastq == fastq_id) %>%
                         dplyr::pull(gene))
    } else {
        valid.genes <- df.meta %>%
            dplyr::filter(fastq == fastq_id) %>%
            dplyr::pull(gene)
    }
    valid.genes <- valid.genes %>% unique()
    df.sum.filt.fwd_rvs <- df.sum.full %>%
        dplyr::mutate(same.gene = sapply(same.gene, function(x){if(is.na(x)){FALSE}else{x}})) %>%
        dplyr::mutate(gene = sapply(1:nrow(.), function(i){
            if (.[i,"same.gene"]) {.[i,"fgene"]}
            else {paste0(.[i,"fgene"], '_', .[i,"rgene"])}
        })) %>%
        dplyr::group_by(gene) %>%
        dplyr::summarise(g.reads = sum(reads)) %>%
        dplyr::ungroup() %>%
        dplyr::rename(reads = g.reads)%>%
        dplyr::mutate(read.type = "fwd_rvs")
    df.sum.filt.fwd <- df.sum.full %>%
        dplyr::select(fwd, rvs, fgene, reads) %>%
        dplyr::distinct() %>%
        dplyr::group_by(fgene) %>%
        dplyr::summarise(reads = sum(reads)) %>%
        dplyr::ungroup() %>%
        dplyr::rename(gene = fgene) %>%
        dplyr::mutate(read.type = "fwd")
    df.sum.filt.rvs <- df.sum.full %>%
        dplyr::select(fwd, rvs, rgene, reads) %>%
        dplyr::distinct() %>%
        dplyr::group_by(rgene) %>%
        dplyr::summarise(reads = sum(reads)) %>%
        dplyr::ungroup() %>%
        dplyr::rename(gene = rgene) %>%
        dplyr::mutate(read.type = "rvs")
    df.sum.filt <- rbind(df.sum.filt.fwd_rvs, df.sum.filt.fwd, df.sum.filt.rvs) %>%
        dplyr::mutate(reads.freq = reads/total.reads,
                      valid.gene = gene %in% valid.genes,
                      fastq = fastq_id) %>%
        dplyr::select(fastq, gene, read.type, reads, reads.freq, valid.gene)
    return(df.sum.filt)
}

summarise_reads_multi <- function(fout, df.meta, dir_sum, dir_blast, valid.genes = c(),
                                  fout_ontarget = NULL){
    ## get samples
    samples <- list.files(dir_blast,
                          pattern = "\\.fwd\\.blastn\\.intersectGene\\.tsv$") %>%
        str_extract("^.+(?=.fwd.blastn.intersectGene.tsv)") %>% unique %>% sort
    ## generate read summary
    df.output <- data.frame()
    for (sample_id in samples){
        print(sample_id)
        df.output <- rbind(df.output,
                           summarise_reads(sample_id, df.meta, dir_sum, dir_blast,
                                           valid.genes = valid.genes) %>%
                           dplyr::arrange(desc(reads)))
    }
    ## write
    write.table(df.output %>%
                dplyr::filter(read.type == "fwd_rvs") %>%
                dplyr::select(-c(read.type)) %>%
                dplyr::mutate(valid.gene = sapply(valid.gene, function(x){ifelse(x, 'T', 'F')})),
                file = fout, quote = FALSE, sep = '\t', row.names = FALSE)
    ## summarise whether each on-target gene is found in output and at what abundance/frequency
    if (!is.null(fout_ontarget)){
        df.filtered <- df.output %>%
            dplyr::select(fastq, gene, read.type, reads, reads.freq) %>%
            ## dplyr::mutate(
            ##            fwd = sapply(gene, function(x){
            ##                if(!str_detect(x, '_')){as.character(x)}else{str_split(x, '_')[[1]][[1]]}}),
            ##            rvs = sapply(gene, function(x){
            ##                if(!str_detect(x, '_')){as.character(x)}else{str_split(x, '_')[[1]][[2]]}})) %>%
            ## dplyr::rename(fwd_rvs = gene) %>%
            ## tidyr::gather(key = "read.type", value = "gene", fwd_rvs, fwd, rvs) %>%
            ## dplyr::group_by(fastq, read.type, gene) %>%
            ## dplyr::summarise(total.reads = sum(reads, na.rm = TRUE),
            ##                  total.reads.freq = sum(reads.freq, na.rm = TRUE)) %>%
            ## dplyr::ungroup()
            dplyr::rename(total.reads = reads, total.reads.freq = reads.freq)
        df.ontarget <- df.meta %>%
            dplyr::select(fastq, gene) %>%
            dplyr::distinct() %>%
            dplyr::left_join(df.filtered, by = c("fastq", "gene"))
        write.table(df.ontarget, file = fout_ontarget, quote = FALSE, sep = '\t', row.name = FALSE)
    }
}

## create output directory
dir_identity <- mkpath(args$dir_output, "identity")
dir.create(dir_identity, showWarnings = FALSE, recursive = TRUE)

## parse metadata from Excel sheet
## df.meta <- parse_meta(args$excel_meta) %>%
##     dplyr::arrange(fastq) %>%
##     dplyr::mutate(gene = str_extract(PCR_Product_fa, args$gene_pattern)) %>%
##     tidyr::separate_rows(gene, sep = ",")
df.meta <- read.table(args$excel_meta, header = TRUE, sep = '\t', stringsAsFactors = FALSE) %>%
    dplyr::arrange(fastq) %>%
    dplyr::mutate(gene = str_extract(PCR_Product_fa, args$gene_pattern)) %>%
    tidyr::separate_rows(gene, sep = ",")

## summarise reads
summarise_reads_multi(mkpath(dir_identity, paste0(prefix, ".identity.txt")),
                      df.meta, args$dir_derep, args$dir_blast,
                      fout_ontarget = mkpath(dir_identity, paste0(prefix, ".identity-ontarget.txt")))

## ## if using primer info
## f_primer_info <- mkpath(dir_proj, "data", "ABE_1st_PCR", "primers_pcr1.tsv")
## valid_genes <- read.table(f_primer_info, header = TRUE, sep = '\t', stringsAsFactors = FALSE) %>%
##     tidyr::separate(amplicon_id, into = c("pool", "gene"), sep = '_') %>%
##     dplyr::pull(gene)
## summarise_reads_multi(mkpath(dir_identity, paste0(prefix, ".identity.txt")),
##                       NA, prefix, dir_sum, dir_blast, valid.genes = valid_genes)
